### PR TITLE
NFDIV-1066: Set up PCQ integration

### PR DESCRIPTION
**Pull request by tomxelliott**
### JIRA link ###
https://tools.hmcts.net/jira/browse/NFDIV-1066

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -97,7 +97,8 @@
     "ts-node": "^10.2.0",
     "tsconfig-paths": "^3.10.1",
     "typescript": "^4.3.5",
-    "uuid": "^8.3.2"
+    "uuid": "^8.3.2",
+    "uuidv4": "^6.2.11"
```
Comment (velimd) [inline package.json:101]: if we have this, do we need both?

**Review by velimd — APPROVED** [review]

File: src/main/steps/applicant1/equality/createToken.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import config from 'config';
+
+const logger = Logger.getLogger('createToken');
+
+const algorithm = 'aes-256-gcm';
+const bufferSize = 16;
+const iv = Buffer.alloc(bufferSize, 0); // Initialization vector.
+const keyLen = 32;
+
+export const createToken = (params: Record<string, unknown>): string => {
```
Comment (linusnorton) [inline src/main/steps/applicant1/equality/createToken.ts:13]: Can we do better than `Record<string, unknown>` here? I'm not really sure what params is supposed to be

File: src/main/steps/applicant1/equality/createToken.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    });
+    const strParams = JSON.stringify(params);
+    const cipher = crypto.createCipheriv(algorithm, key, iv);
+    encrypted = cipher.update(strParams, 'utf8', 'hex');
+    encrypted += cipher.final('hex');
+  } else {
+    logger.error('PCQ token key is missing.');
+  }
+
+  return encrypted;
```
Comment (linusnorton) [inline src/main/steps/applicant1/equality/createToken.ts:31]: I'm not quite sure I like the way this method can silently fail, how does the calling code deal with that?

File: src/main/steps/applicant1/equality/createToken.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+const algorithm = 'aes-256-gcm';
+const bufferSize = 16;
+const iv = Buffer.alloc(bufferSize, 0); // Initialization vector.
+const keyLen = 32;
+
+export const createToken = (params: Record<string, unknown>): string => {
+  const tokenKey: string = config.get('services.equalityAndDiversity.tokenKey');
+  let encrypted = '';
+
+  if (tokenKey) {
```
Comment (linusnorton) [inline src/main/steps/applicant1/equality/createToken.ts:17]: do you need this check? is there ever a situation where the tokenKey is not set?

File: src/main/steps/applicant1/equality/createToken.ts (hunk 3 above)
Comment (tomxelliott) [reply src/main/steps/applicant1/equality/createToken.ts:31]: If the token is returned as empty or there are any issues the user will see the PCQ unavailable page

File: src/main/steps/applicant1/equality/createToken.ts (hunk 2 above)
Comment (tomxelliott) [reply src/main/steps/applicant1/equality/createToken.ts:13]: Yeah can create a PCQ params type or something like that

File: src/main/steps/applicant1/equality/createToken.ts (hunk 4 above)
Comment (tomxelliott) [reply src/main/steps/applicant1/equality/createToken.ts:17]: Only if there is an issue getting it from the config but I am happy to remove this conditional

**Pull request merged**